docs/data-model/zng.md

```diff
@@ -61,69 +56,54 @@ or otherwise rewritten to be merged in this fashion.
 
 ## 2. The ZNG Format
 
-A ZNG stream comprises a sequence of interleaved control messages and value messages
-that are serialized into a stream of bytes.
-
-Each message is prefixed with a single-byte header code.  Codes `0xf5-0xff`
-are allocated as control messages while codes `0x00-0xf5` indicate a value message.
-
-### 2.1 Control Messages
-
-Control codes `0xf5` through `0xff` (in hexadecimal) are defined as follows:
-
-| Code   | Message Type                   |
-|--------|--------------------------------|
-| `0xf5` | record type definition         |
-| `0xf6` | array type definition          |
-| `0xf7` | set type definition            |
-| `0xf8` | map type definiton             |
-| `0xf9` | union type definition          |
-| `0xfa` | enum type definiton            |
-| `0xfb` | named type definition          |
-| `0xfc` | error type definiton           |
-| `0xfd` | compressed value message block |
-| `0xfe` | application-defined message    |
-| `0xff` | end-of-stream                  |
-
-The application-defined messages are available to higher-layer protocols and
-potential future variations of ZNG.  A ZNG implementation that
-merely skips over all of the application-defined messages is guaranteed by
-this specification to decode all of the data as described herein even if such
-messages provide additional semantics on top of the base ZNG format.
+A ZNG stream comprises a sequence of interleaved frames of data
+where a frame contains one of three types of data:
+_types_, _values_, or externally-defined _control_.
 
-Any such application-defined messages not known by
-a ZNG data receiver shall be ignored.
+A stream is punctuated by the end-of-stream value `0xff`.
 
-The body of a application-defined control message is typically a structured
-message in JSON, ZSON, or ZNG.
-These messages are guaranteed to be preserved
-in order within the stream and presented to higher layer components through
-any ZNG streaming API.  In this way, senders and receivers of ZNG can embed
-protocol directives as ZNG control payloads rather than defining additional
-encapsulating protocols.
+Each frame header includes a length field
+allowing an implementation to easily skip from frame to frame.
 
-> For example, the [Zed service](../../docs/lake/service-api.md) query endpoint
-> uses application-defined message `0xfe` to embed search and server stats in
-> the return stream of ZNG data, e.g., as a long-running search progresses on
-> the server.
+Each frame begins with its version bit and a code indicating
+whether it is compressed or not:
+
+|    code    |  len    | body            |
+|------------|---------|-----------------|
+| `00xxxxxx` | uvarint | uncompressed data |
+| `01xxxxxx` | uvarint | compressed data |
```

Comment:
Maybe add an RFC-style figure describing the code byte, along these lines?
```
    7 6 5 4 5 3 1 0 
   +-+-+-+-+-+-+-+-+
   |V|C|  T|      L|
   +-+-+-+-+-+-+-+-+

   V: 1 bit

     Version number.  Must be zero.

   C: 1 bit

     Indicates compressed frame data.

   T: 2 bits

     Type of frame data.

       00: Types
       01: Values
       10: Control
       11: End of stream

   L: 4 bits

     High-order bits of frame length.
```

Reply:
I found it easier to write and follow while I was coding to unravel the various cases but I will write it up this way.

Reply:
We can do this later.